README.md
# EHDEV-CONFIGER-SCMS
---

[![Build Status](https://travis-ci.org/EHDFE/ehdev-configer-scms.svg?branch=master)](https://travis-ci.org/EHDFE/ehdev-configer-scms)
[![npm](https://img.shields.io/npm/dm/ehdev-configer-scms.svg)]()
[![npm](https://img.shields.io/npm/v/ehdev-configer-scms.svg)]()
[![node](https://img.shields.io/node/v/ehdev-configer-scms.svg)]()
[![GitHub tag](https://img.shields.io/github/tag/ehdfe/ehdev-configer-scms.svg)]()
[![David](https://img.shields.io/david/EHDFE/ehdev-configer-scms.svg)]()
[![David](https://img.shields.io/david/dev/EHDFE/ehdev-configer-scms.svg)]()


## 配置说明

| 配置项 | 默认值 | 说明 |
|---|---| ---|
| buildPath | dist | 输出目录 |
| enableHotModuleReplacement | true | 启用[热更新](https://webpack.js.org/guides/hot-module-replacement) |
| ignoreHtmlTemplate | false | 忽略输出 html |
| entryList | `{app: './index'}` | 配置 entry |
| htmlList | `[]` | 用于 htmlWebpackPlugin 配置，如 `[{"files": {"js": ["index.less"]},"chunks": ["app"],"filename":"index.html","template": "./index.html"}` |
|htmlWebpackPlugin|`{inject: true, chunksSortMode: 'auto', cache: true, showErrors: true}`|`htmlWebpackPlugin` 插件配置, 参考 [https://github.com/jantimon/html-webpack-plugin#configuration](https://github.com/jantimon/html-webpack-plugin#configuration)|
| alias | `{}` | 对应 webpack 配置 [resolve.alias](https://webpack.js.org/configuration/resolve/#resolve-alias) |
| externals | `{}` | 对应 webpack 配置 [externals](https://webpack.js.org/configuration/externals/#externals) |
| browserSupports | last 2 version | 浏览器支持配置，影响 `babel` 和 `autoprefixer`, 配置参考：[https://github.com/ai/browserslist](https://github.com/ai/browserslist) |
| dll | `{ enable: false, enclude: [] }` | 是否启用 dll，enclude 提供打入 dll 包的模块 |
| providePluginConfig | `{}` | 主要用来支持 jQuery 依赖全局挂载的老模块， 参考 [https://webpack.js.org/plugins/provide-plugin/](https://webpack.js.org/plugins/provide-plugin/) |
| babelUseBuiltIns | true | [babel-preset-env#usebuiltins](http://babeljs.io/docs/plugins/preset-env/#usebuiltins) 配置 |
| https | false | 开发环境的 https 支持 |
| publicPath | `../` | `webpackConfig.output.publicPath`, 只在构建时生效 |
| supportIE8 | false | 支持 ie8 |
| contentBase | undefined | 配置 devServer 的 [contentBase](https://webpack.js.org/configuration/dev-server/#devserver-contentbase)，默认包含当前项目的输出目录，不需要配置 |


HISTORY.md
# HISTORY

## v0.0.1
---
- init

index.js
/**
 * merge and export
 * @author ryan.bian
 */
const merge = require('webpack-merge');

const _getBaseConfig = require('./src/base.conf');
const _getDevConfig = require('./src/dev.conf');
const _getProdConfig = require('./src/prod.conf');

const DEFAULT_PROJECT_CONFIG = require('./src/projectConfig');
const MERGE_STRATEGY = {
  plugins: 'append',
};

exports.DEFAULT_PROJECT_CONFIG = DEFAULT_PROJECT_CONFIG;

exports.getDevConfig = async (PROJECT_CONFIG, options) => {
  try {
    const baseConfig = await _getBaseConfig(PROJECT_CONFIG, options);
    const devConfig = await _getDevConfig(PROJECT_CONFIG, options);
    return merge.smartStrategy(MERGE_STRATEGY)(baseConfig, devConfig);
  } catch(e) {
    throw Error(e);
  }
};

exports.getProdConfig = async (PROJECT_CONFIG, options) => {
  try {
    const baseConfig = await _getBaseConfig(PROJECT_CONFIG, options);
    const prodConfig = await _getProdConfig(PROJECT_CONFIG, options);
    return merge.smartStrategy(MERGE_STRATEGY)(baseConfig, prodConfig);
  } catch(e) {
    throw Error(e);
  }
};

src/lib.js
const path = require('path');
const fs = require('fs');

const SHELL_NODE_MODULES_PATH = process.env.SHELL_NODE_MODULES_PATH;
const webpack = require(path.join(SHELL_NODE_MODULES_PATH, 'webpack'));

const PROJECT_ROOT = exports.PROJECT_ROOT = process.cwd();
const SOURCE_DIR = exports.SOURCE_DIR = path.join(PROJECT_ROOT, 'src');

exports.getExternals = PROJECT_CONFIG => {
  let externals = Object.assign({}, PROJECT_CONFIG.externals);
  if (PROJECT_CONFIG.externalModulesMap) {
    const externalModuleNames = Object.keys(PROJECT_CONFIG.externalModulesMap);
    externals = [
      externals,
      (context, request, callback) => {
        const matchName = externalModuleNames.find(name => request.startsWith(name));
        if (matchName) {
          const restPath = request.replace(`${matchName}/`, '');
          return callback(null, `${PROJECT_CONFIG.externalModulesMap[matchName]}('${restPath}')`)
        }
        callback();
      },
    ]
  }
  return externals;
};

exports.getHtmlLoaderConfig = PROJECT_CONFIG => ({
  test: /\.html?$/,
  use: [
    {
      loader: require.resolve('html-loader'),
      options: {
        ignoreCustomFragments: [/\{\{.*?}}/],
        interpolate: true,
        root: './',
      },
    },
    {
      loader: require.resolve('posthtml-loader'),
    },
  ],
});

exports.getLoaderOptionPlugin = PROJECT_CONFIG => new webpack.LoaderOptionsPlugin({
  options: {
    posthtml(ctx) {
      return {
        plugins: [
          require('posthtml-expressions')({
            locals:{
              env: process.env.NODE_ENV, 
            },
            // defaults delimiters: {{}} is conflicted with angular expression
            delimiters: ['<%', '%>'],
          })
        ],
      };
    }
  },
});

src/prod.conf.js
/**
 * development config
 */
const path = require('path');
const SHELL_NODE_MODULES_PATH = process.env.SHELL_NODE_MODULES_PATH;
const webpack = require(path.join(SHELL_NODE_MODULES_PATH, 'webpack'));
const HtmlWebpackPlugin = require(path.join(SHELL_NODE_MODULES_PATH, 'html-webpack-plugin'));
const ExtractTextPlugin = require(path.join(SHELL_NODE_MODULES_PATH, 'extract-text-webpack-plugin'));
const CleanWebpackPlugin = require(path.join(SHELL_NODE_MODULES_PATH, 'clean-webpack-plugin'));
const autoprefixer = require('autoprefixer');
const WebpackChunkHash = require('webpack-chunk-hash');
const ManifestPlugin = require('webpack-manifest-plugin');

const {
  PROJECT_ROOT,
  SOURCE_DIR,
  getExternals,
  getHtmlLoaderConfig,
  getLoaderOptionPlugin,
} = require('./lib');
const PUBLIC_PATH = '/';

module.exports = async (PROJECT_CONFIG, options) => {

  const configResult = {};

  const entry = PROJECT_CONFIG.entryList;

  const output = {
    path: path.join(PROJECT_ROOT, PROJECT_CONFIG.buildPath),
    filename: '[name].[chunkhash:8].js',
    publicPath: PROJECT_CONFIG.publicPath || PUBLIC_PATH,
  };

  if (PROJECT_CONFIG.outputJsonpFunction) {
    Object.assign(output, {
      jsonpFunction: PROJECT_CONFIG.outputJsonpFunction
    });
  }

  const babelLoaderConfig = {
    loader: require.resolve('babel-loader'),
    options: {
      // @remove-on-eject-begin
      babelrc: false,
      presets: [
        [
          require.resolve('babel-preset-env'),
          {
            targets: {
              browsers: PROJECT_CONFIG.browserSupports.PRODUCTION,
            }, 
            module: false,
            useBuiltIns: PROJECT_CONFIG.babelUseBuiltIns,
          }
        ],
        require.resolve('babel-preset-stage-1'),
      ],
      // @remove-on-eject-end
      // This is a feature of `babel-loader` for webpack (not Babel itself).
      // It enables caching results in ./node_modules/.cache/babel-loader/
      // directory for faster rebuilds.
      cacheDirectory: true,
    },
  }

  const module = {
    rules: [
      {
        // "oneOf" will traverse all following loaders until one will
        // match the requirements. When no loader matches it will fall
        // back to the "file" loader at the end of the loader list.
        oneOf: [
          // "url" loader works like "file" loader except that it embeds assets
          // smaller than specified limit in bytes as data URLs to avoid requests.
          // A missing `test` is equivalent to a match.
          {
            test: [/\.bmp$/, /\.gif$/, /\.jpe?g$/, /\.png$/],
            loader: require.resolve('url-loader'),
            options: {
              limit: 10000,
              name: '[name].[hash:8].[ext]',
            },
          },
          // Process JS with Babel.
          Object.assign({
            test: /\.js$/,
            include: SOURCE_DIR,
          }, babelLoaderConfig),
          {
            test: /\.(le|c)ss$/,
            loader: ExtractTextPlugin.extract({
              fallback: {
                loader: require.resolve('style-loader'),
                options: {
                  hmr: false,
                },
              },
              use: [
                {
                  loader: require.resolve('css-loader'),
                  options: {
                    importLoaders: 1,
                    minimize: false,
                  },
                },
                {
                  loader: require.resolve('postcss-loader'),
                  options: {
                    // Necessary for external CSS imports to work
                    // https://github.com/facebookincubator/create-react-app/issues/2677
                    ident: 'postcss',
                    plugins: () => [
                      autoprefixer({
                        browsers: PROJECT_CONFIG.browserSupports.DEVELOPMENT,
                      }),
                    ],
                  },
                },
                {
                  loader: require.resolve('less-loader'),
                },
              ],
            }),
          },
          getHtmlLoaderConfig(PROJECT_CONFIG),
          // "file" loader makes sure those assets get served by WebpackDevServer.
          // When you `import` an asset, you get its (virtual) filename.
          // In production, they would get copied to the `build` folder.
          // This loader doesn't use a "test" so it will catch all modules
          // that fall through the other loaders.
          {
            // Exclude `js` files to keep "css" loader working as it injects
            // it's runtime that would otherwise processed through "file" loader.
            // Also exclude `html` and `json` extensions so they get processed
            // by webpacks internal loaders.
            exclude: [/\.jsx?$/, /\.json$/],
            loader: require.resolve('file-loader'),
            options: {
              name: '[name].[hash:8].[ext]',
            },
          },
        ],
      }
    ]
  };

  const resolve = {
    alias: PROJECT_CONFIG.alias,
  };

  const externals = getExternals(PROJECT_CONFIG);

  const plugins = [
    new CleanWebpackPlugin([
      PROJECT_CONFIG.buildPath,
    ], {
      root: PROJECT_ROOT,
      verbose: true,
      dry: false,
    }),
    new webpack.HashedModuleIdsPlugin(),
    new WebpackChunkHash(),
    new ExtractTextPlugin({
      filename: '[name].[contenthash:8].css',
    }),
    new ManifestPlugin(),
    getLoaderOptionPlugin(PROJECT_CONFIG),
  ];
  if (!PROJECT_CONFIG.ignoreHtmlTemplate) {
    PROJECT_CONFIG.htmlList.forEach(d => {
      plugins.push(
        new HtmlWebpackPlugin(Object.assign({
          filename: d.filename,
          template: d.template,
          files: d.files,
          chunks: d.chunks,
        }, PROJECT_CONFIG.HtmlWebpackPlugin)),
      );
    });
  }

  Object.assign(configResult, {
    // Don't attempt to continue if there are any errors.
    bail: true,
    entry,
    output,
    module,
    resolve,
    externals,
    plugins,
    devtool: 'source-map',
  });

  return configResult;
};

src/projectConfig.js
module.exports = {
  buildPath: 'dist',
  enableHotModuleReplacement: true,
  ignoreHtmlTemplate: false,
  entryList: {
    app: './index',
  },
  htmlList: [],
  htmlWebpackPlugin: {
    // true | 'head' | 'body' | false Inject all assets into the given template or templateContent
    // When passing true or 'body' all javascript resources will be placed at the bottom of the body element.
    // 'head' will place the scripts in the head element.
    inject: true,
    // Allows to control how chunks should be sorted before they are included to the html.
    // Allowed values: 'none' | 'auto' | 'dependency' |'manual' | {function}
    chunksSortMode: 'auto',
    // if true (default) try to emit the file only if it was changed.
    cache: true,
    // if true (default) errors details will be written into the HTML page.
    showErrors: true,
  },
  alias: {},
  externals: {},
  browserSupports: {
    DEVELOPMENT: ['last 2 versions'],
    PRODUCTION: ['last 2 versions'],
  },
  dll: {
    enable: false,
    include: [],
  },
  providePluginConfig: {},
  babelUseBuiltIns: true,
  // valid in dev environment
  https: false,
  publicPath: '../',
  // outputJsonpFunction
  // contentBase
};

src/dev.conf.js
/**
 * development config
 */
const path = require('path');
const SHELL_NODE_MODULES_PATH = process.env.SHELL_NODE_MODULES_PATH;
const webpack = require(path.join(SHELL_NODE_MODULES_PATH, 'webpack'));
const HtmlWebpackPlugin = require(path.join(SHELL_NODE_MODULES_PATH, 'html-webpack-plugin'));
const ManifestPlugin = require('webpack-manifest-plugin');
const autoprefixer = require('autoprefixer');

const {
  PROJECT_ROOT,
  SOURCE_DIR,
  getExternals,
  getHtmlLoaderConfig,
  getLoaderOptionPlugin,
} = require('./lib');

module.exports = async (PROJECT_CONFIG, options) => {

  let PUBLIC_PATH = '/';
  if (options.ip) {
    PUBLIC_PATH = `${PROJECT_CONFIG.https ? 'https' : 'http'}://${options.ip}:${options.port}/`;
  }

  const configResult = {};

  const entry = PROJECT_CONFIG.entryList;

  if (PROJECT_CONFIG.enableHotModuleReplacement) {
    const devServerEntry = [
      `${require.resolve(`${path.join(SHELL_NODE_MODULES_PATH, 'webpack-dev-server')}/client`)}?http://localhost:${options.port}`,
      require.resolve(`${path.join(SHELL_NODE_MODULES_PATH, 'webpack')}/hot/dev-server`),
    ];
    Object.keys(entry).forEach(name => {
      let entryValue = entry[name];
      if (Array.isArray(entryValue)) {
        entryValue = entryValue.unshift(...devServerEntry);
      } else {
        entryValue = [...devServerEntry, entryValue];
      }
      Object.assign(entry, {
        [name]: entryValue,
      });
    });
  }

  const output = {
    path: path.join(PROJECT_ROOT, PROJECT_CONFIG.buildPath),
    // Add /* filename */ comments to generated require()s in the output.
    pathinfo: true,
    publicPath: PUBLIC_PATH,
  };

  if (PROJECT_CONFIG.outputJsonpFunction) {
    Object.assign(output, {
      jsonpFunction: PROJECT_CONFIG.outputJsonpFunction
    });
  }

  const babelLoaderConfig = {
    loader: require.resolve('babel-loader'),
    options: {
      // @remove-on-eject-begin
      babelrc: false,
      presets: [
        [
          require.resolve('babel-preset-env'),
          {
            targets: {
              browsers: PROJECT_CONFIG.browserSupports.DEVELOPMENT,
            }, 
            module: false,
            useBuiltIns: PROJECT_CONFIG.babelUseBuiltIns,
          }
        ],
        require.resolve('babel-preset-stage-1'),
      ],
      // @remove-on-eject-end
      // This is a feature of `babel-loader` for webpack (not Babel itself).
      // It enables caching results in ./node_modules/.cache/babel-loader/
      // directory for faster rebuilds.
      cacheDirectory: true,
    },
  }

  const module = {
    rules: [
      {
        // "oneOf" will traverse all following loaders until one will
        // match the requirements. When no loader matches it will fall
        // back to the "file" loader at the end of the loader list.
        oneOf: [
          // "url" loader works like "file" loader except that it embeds assets
          // smaller than specified limit in bytes as data URLs to avoid requests.
          // A missing `test` is equivalent to a match.
          {
            test: [/\.bmp$/, /\.gif$/, /\.jpe?g$/, /\.png$/],
            loader: require.resolve('url-loader'),
            options: {
              limit: 10000,
              name: '[name].[hash:8].[ext]',
            },
          },
          // Process JS with Babel.
          Object.assign({
            test: /\.js$/,
            include: SOURCE_DIR,
          }, babelLoaderConfig),
          {
            test: /\.(le|c)ss$/,
            use: [
              require.resolve('style-loader'),
              {
                loader: require.resolve('css-loader'),
                options: {
                  importLoaders: 1,
                  minimize: false,
                },
              },
              {
                loader: require.resolve('postcss-loader'),
                options: {
                  // Necessary for external CSS imports to work
                  // https://github.com/facebookincubator/create-react-app/issues/2677
                  ident: 'postcss',
                  plugins: () => [
                    autoprefixer({
                      browsers: PROJECT_CONFIG.browserSupports.DEVELOPMENT,
                    }),
                  ],
                },
              },
              {
                loader: require.resolve('less-loader'),
              }
            ],
          },
          getHtmlLoaderConfig(PROJECT_CONFIG),
          // "file" loader makes sure those assets get served by WebpackDevServer.
          // When you `import` an asset, you get its (virtual) filename.
          // In production, they would get copied to the `build` folder.
          // This loader doesn't use a "test" so it will catch all modules
          // that fall through the other loaders.
          {
            // Exclude `js` files to keep "css" loader working as it injects
            // it's runtime that would otherwise processed through "file" loader.
            // Also exclude `html` and `json` extensions so they get processed
            // by webpacks internal loaders.
            exclude: [/\.jsx?$/, /\.json$/],
            loader: require.resolve('file-loader'),
            options: {
              name: '[name].[hash:8].[ext]',
            },
          },
        ],
      }
    ]
  };

  const resolve = {
    alias: PROJECT_CONFIG.alias,
  };

  const externals = getExternals(PROJECT_CONFIG);

  const plugins = [
    // Add module names to factory functions so they appear in browser profiler.
    new webpack.NamedModulesPlugin(),
    new ManifestPlugin(),
    getLoaderOptionPlugin(PROJECT_CONFIG),
  ];
  if (PROJECT_CONFIG.enableHotModuleReplacement) {
    plugins.push(
      // This is necessary to emit hot updates (currently CSS only):
      new webpack.HotModuleReplacementPlugin(),
    );
  }
  if (!PROJECT_CONFIG.ignoreHtmlTemplate) {
    PROJECT_CONFIG.htmlList.forEach(d => {
      plugins.push(
        new HtmlWebpackPlugin(Object.assign({
          filename: d.filename,
          template: d.template,
          files: d.files,
          chunks: d.chunks,
        }, PROJECT_CONFIG.HtmlWebpackPlugin)),
      );
    });
  }

  Object.assign(configResult, {
    entry,
    output,
    module,
    resolve,
    externals,
    plugins,
    devtool: 'cheap-module-source-map',
    // Turn off performance hints during development because we don't do any
    // splitting or minification in interest of speed. These warnings become
    // cumbersome.
    performance: {
      hints: false,
    },
  });

  return configResult;
};